### ISSUE TITLE
Firefox WebExtensions

### ISSUE THREAD

**Issue by Mottie** [JasonBarnabe/stylish#247]
I don't know if you know, but Firefox is planning to standardize their addons to match Webkit by introducing a new [WebExtensions API](https://blog.mozilla.org/addons/2015/08/21/the-future-of-developing-firefox-add-ons/).

> A preview release of WebExtensions is available in Firefox 42, which is currently on Developer Edition...

I was wondering if you had any plans on changing this Firefox addon, and if so, do you have any ideas on a timeline?

The reason I am asking is because a user was requesting use to add a button to toggle an applied Stylish theme, so I created [this userscript](https://github.com/StylishThemes/Stylish-Toggle) which, so far, works in Chrome & Opera because the style is injected into the page. It doesn't work in Firefox because the style isn't injected. So, I was just curious :grin:

**Comment by theres-waldo** [comment]
Note that Mozilla has recently [announced](https://blog.mozilla.org/addons/2016/11/23/add-ons-in-2017/) that starting in Firefox 57, only WebExtensions will be supported.

Are there plans to convert Stylish into a WebExtension?

**Comment by jscher2000** [comment]
> Are there plans to convert Stylish into a WebExtension?

As of last June: "the current plan is to toss all of this and use the code from Stylish for Chrome." See: https://github.com/stylish-userstyles/stylish/issues/275#issuecomment-225079880

But since then, we have new owners, so... who knows whether that will happen in a timely manner? And of course it means no ability to style chrome.

**Comment by BenoitZugmeyer** [comment]
Non-webextensions are completely disabled with the latest Nightly. So I tried to install the Chrome stylish extension, and it seems to work mostly out of the box! That's pretty nice. I think supporting Firefox wouldn't be too complicated.

**Comment by mikhoul** [comment]
@BenoitZugmeyer   Hi just use this addon it's fully compatible and maintained and 100% webext, moreover it's better than Stylish 😸  

Link: https://github.com/FirefoxBar/xStyle/releases

Regards :octocat:

**Comment by BenoitZugmeyer** [comment]
@mikhoul Thanks for sharing, this is great!

**Comment by JZersche** [comment]
xStyle doesn't parse my code the same as Stylish, and doesn't work with Browser chrome files. I think Stylish is better.

**Comment by Mottie** [comment]
I would recommend using Stylus - https://github.com/openstyles/stylus

**Comment by JZersche** [comment]
Stylus doesn't work with Browser chrome files, and has similar issues parsing my code. I just reverted to 52.0b9. Now everything works properly again, yay.


![100](https://user-images.githubusercontent.com/24817874/29622894-a58dcea4-87f2-11e7-8f84-16c13d20987f.jpg)

**Comment by jscher2000** [comment]
Hi  JZersche -

> xStyle ... doesn't work with Browser chrome files.

> Stylus doesn't work with Browser chrome files...

You can generalize that to "WebExtensions cannot style browser chrome" (unless/until a new theming API come to fruition, and even then one should not expect the same freedom offered in the past).

Welcome back to the era of userChrome.css files.

> I just reverted to 52.0b9. Now everything works properly again, yay.

An old beta? If you want to use the Firefox 52 branch, try the Extended Support Release, which will continue to receive security updates through next April or so. https://www.mozilla.org/firefox/organizations/all/

**Comment by mikhoul** [comment]
Hi @JZersche  like @jscher2000 said it's not a limitation of xStyle but a limitation of Mozilla/Firefox by not providing the adequate API to apply CSS to the chrome of Firefox for addon.

I've also reverted to using "userChrome.css" to modify the Chrome of Firefox. 🤕 

**BTW:** The style editor of xStyle in one of the best I've seen with Userstyle managers, moreover the dev is very open to suggestions to improve xStyle.  Right now I'm working on a french translation 🌍 .

**N.B.:**  xStyle is now available on AMO https://addons.mozilla.org/zh-CN/firefox/addon/xstyle/  but to have the latest signed version it's better to download it from the GitHub repo here: https://github.com/FirefoxBar/xStyle/releases  Mozilla take longtime to review addon so there is lot of delay sometime between the submission of an addon on AMO and the availability of the addon.

Stylus is good too but it's a fork of the old code from the Chrome Stylish and xStyle is developed first for Firefox and ported/tested to Chrome after. 

Regards :octocat:

**Comment by Mottie** [comment]
@mikhoul xStyle was also forked from Chrome Stylish ([evidence](https://github.com/FirefoxBar/xStyle/commit/78bb18fa5b36d6026e53dfc89bc590419bfcbc65))... I haven't had a chance to check it out.

Update: I tried out the extension and there are still a bunch of issues with it, especially with our [GitHub-Dark](https://github.com/StylishThemes/GitHub-Dark) style. I do like the color picker; but even that is a bit problematic:

![](https://user-images.githubusercontent.com/136959/29590439-70527f98-875f-11e7-8c9c-9df33b5b5a31.png)

**Comment by mikhoul** [comment]
@Mottie if you see bugs/conflicts with xStyle open an issue and there will be fixed: https://github.com/FirefoxBar/xStyle/issues/new

Regards :octocat:

**Comment by JZersche** [comment]
It doesn't seem like they have API's with WebExtensions to create new toolbars either. I noticed WebExtensions not only broke Stylish but also the 'The Add-On Bar Restored' extension. Hopefully they.. add additional functionality to WebExtensions. Here's a userstyles.org theme I just made if anywould would like to use it.

https://userstyles.org/styles/147064/userstyles-org-zdark-skin

**Comment by JohnDDuncanIII** [comment]
So, the webextensions API does not support any internal chrome:// theme-ing, and it is not possible to style scrollbars using the userChrome.css (the only thing I used stylish for). Webextensions seem to severely limit the creative potential and power of old XUL/XPCOM-based extensions; Mozilla is really leaving power users out to dry with this one.

**Comment by stuaxo** [comment]
It's probably worth contacting them and asking for APIs for what you need.
